File: CodingInterviewsJavaSolutions/48. Rotate Image.java
class Solution {
    public void rotate(int[][] matrix) {
        int n = matrix.length;

        for (int r = 0; r < n; r++) {
            for (int c = r + 1; c < n; c++) {
                int temp = matrix[r][c];
                matrix[r][c] = matrix[c][r];
                matrix[c][r] = temp;
            }
        }

        for (int r = 0; r < n; r++) {
            for (int c = 0; c < n / 2; c++) {
                int temp = matrix[r][c];
                matrix[r][c] = matrix[r][n - c - 1];
                matrix[r][n - c - 1] = temp;
            }
        }
    }
}

File: CodingInterviewsPackSolutionsCPP/48. Rotate Image.cpp
class Solution {
public:
    void transpose(vector<vector<int>>&matrix){
        int n = matrix.size();

        for(int r = 0; r < n; r++){
            for(int c = r+1; c < n; c++){
                swap(matrix[r][c],matrix[c][r]);
            }
        }
    }

    void reflect(vector<vector<int>>&matrix){
        int n = matrix.size();
        for(int r = 0; r < n/2; r++){
            for(int c = 0; c < n; c++){
                swap(matrix[c][r],matrix[c][n-r-1]);
            }
        }
    }

    void rotate(vector<vector<int>>& matrix) {
        transpose(matrix);
        reflect(matrix);
    }
};